src/rustc.rs

```diff
@@ -79,6 +81,255 @@ pub fn hash_from_version_string(version: &str, index: usize) -> String {
+        this
+    }
+
+    /// Replace the host, does nothing if ran on a custom toolchain
+    pub fn replace_host(&mut self, host: &ImagePlatform) -> &mut Self {
+        if !self.is_custom {
+            *self = Self::new(&self.channel, &self.date, host, &self.sysroot);
+            self.sysroot.set_file_name(&self.full);
+        }
+        self
+    }
+
+    /// Makes a good guess as to what the toolchain is compiled to run on.
+    pub(crate) fn custom(
+        name: &str,
+        sysroot: &Path,
+        config: &crate::config::Config,
+        msg_info: &mut MessageInfo,
+    ) -> Result<QualifiedToolchain> {
+        if let Some(compat) = config.custom_toolchain_compat() {
+            let mut toolchain: QualifiedToolchain = QualifiedToolchain::parse(
+                sysroot.to_owned(),
+                &compat,
+                config,
+                msg_info,
+            )
+            .wrap_err(
+                "could not parse CROSS_CUSTOM_TOOLCHAIN_COMPAT as a fully qualified toolchain name",
+            )?;
+            toolchain.is_custom = true;
+            toolchain.full = name.to_owned();
+            return Ok(toolchain);
+        }
+        // a toolchain installed by https://github.com/rust-lang/cargo-bisect-rustc
+        if name.starts_with("bisector-nightly") {
+            let (_, toolchain) = name.split_once('-').expect("should include -");
+            let mut toolchain =
+                QualifiedToolchain::parse(sysroot.to_owned(), toolchain, config, msg_info)
+                    .wrap_err("could not parse bisector toolchain")?;
+            toolchain.is_custom = true;
+            toolchain.full = name.to_owned();
+            return Ok(toolchain);
+        } else if let Ok(stdout) = Command::new(sysroot.join("bin/rustc"))
+            .arg("-Vv")
+            .run_and_get_stdout(msg_info)
+        {
+            let rustc_version::VersionMeta {
+                build_date,
+                channel,
+                host,
+                ..
+            } = rustc_version::version_meta_for(&stdout)?;
+            let mut toolchain = QualifiedToolchain::new(
+                match channel {
+                    rustc_version::Channel::Dev => "dev",
+                    rustc_version::Channel::Nightly => "nightly",
+                    rustc_version::Channel::Beta => "beta",
+                    rustc_version::Channel::Stable => "stable",
+                },
+                &build_date,
+                &ImagePlatform::from_target(host.into())?,
+                sysroot,
+            );
+            toolchain.is_custom = true;
+            toolchain.full = name.to_owned();
+            return Ok(toolchain);
+        }
+        Err(eyre::eyre!(
+            "cross can not figure out what your custom toolchain is"
+        ))
+        .suggestion("set `CROSS_CUSTOM_TOOLCHAIN_COMPAT` to a fully qualified toolchain name: i.e `nightly-aarch64-unknown-linux-musl`")
```

Comment:
`CUSTOM_TOOLCHAIN_COMPAT` should probably be documented, maybe on the wiki. I'll add this to my TODO list.